### PR TITLE
Refactor writable warm feature flag to include tiering

### LINKED ISSUE #13718: Refactor the writable warm index feature flag
Refactor the writable remote index feature flag to include the tiering functionality as well. Considered the following options:

1. TIERED_REMOTE_INDEX
2. TIERED_WRITEABLE_INDEX
3. TIERED_WRITEABLE_REMOTE_INDEX (Preferred)
4. TIERED_WRITEABLE_WARM_INDEX
5. TIERING_EXPERIMENTAL

### PR DESCRIPTION
**Pull request by neetikasinghal**
### Description
Refactor the writable remote index feature flag to include the tiering functionality as well. Considered the following options:

1. TIERED_REMOTE_INDEX
2. TIERED_WRITEABLE_INDEX
3. TIERED_WRITEABLE_REMOTE_INDEX (Preferred)
4. TIERED_WRITEABLE_WARM_INDEX
5. TIERING_EXPERIMENTAL

### Related Issues
Resolves #13718 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/settings/FeatureFlagSettings.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ protected FeatureFlagSettings(
         FeatureFlags.IDENTITY_SETTING,
         FeatureFlags.TELEMETRY_SETTING,
         FeatureFlags.DATETIME_FORMATTER_CACHING_SETTING,
-        FeatureFlags.WRITEABLE_REMOTE_INDEX_SETTING,
+        FeatureFlags.TIERED_WRITEABLE_REMOTE_INDEX_SETTING,
```
Comment (sohami) [inline server/src/main/java/org/opensearch/common/settings/FeatureFlagSettings.java:37]: Can we use `TIERED_REMOTE_STORE_INDEX_SETTING` ? This covers both tiering and remote store index aspect. By default, remote store backed index supports read/write so we don't have to really mention that in feature flag.

Comment (neetikasinghal) [reply server/src/main/java/org/opensearch/common/settings/FeatureFlagSettings.java:37]: In that case, I think `TIERED_REMOTE_INDEX_SETTING` should be sufficient. STORE also is redundant, let me know if that sounds good to u as well

**Comment by neetikasinghal** [comment]
> ❌ Gradle check result for [6e9632c](https://github.com/opensearch-project/OpenSearch/commit/6e9632c8b701a0df4880c637ee7779601554ef65): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/38943/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/13473

File: server/src/main/java/org/opensearch/common/settings/FeatureFlagSettings.java (hunk 1 above)
Comment (sohami) [reply server/src/main/java/org/opensearch/common/settings/FeatureFlagSettings.java]: I am fine with remote Index as well. To me remote store index was more explicit in terms of index store is remote vs index is remote (which is not the case here).

File: server/src/main/java/org/opensearch/common/util/FeatureFlags.java [hunk 2]
```diff
@@ -56,10 +56,10 @@ public class FeatureFlags {
     public static final String DATETIME_FORMATTER_CACHING = "opensearch.experimental.optimization.datetime_formatter_caching.enabled";
 
     /**
-     * Gates the functionality of writeable remote index
+     * Gates the functionality of writeable remote index having the capability to move across different tiers
```
Comment (sohami) [inline server/src/main/java/org/opensearch/common/util/FeatureFlags.java:59]: `.. functionality of writeable remote index..` ---> `.. functionality of remote index...`

**Comment by neetikasinghal** [comment]
> ❌ Gradle check result for [756e26a](https://github.com/opensearch-project/OpenSearch/commit/756e26acdd33f85b9c203dbcbd366fa01a308af8): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39104/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/12197
https://github.com/opensearch-project/OpenSearch/issues/13600

**Comment by sohami** [comment]
> > ❌ Gradle check result for [756e26a](https://github.com/opensearch-project/OpenSearch/commit/756e26acdd33f85b9c203dbcbd366fa01a308af8): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39104/)
> > Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
> 
> #12197 #13600

Created tracking issue for other variants of failure as well: https://github.com/opensearch-project/OpenSearch/issues/13792

**Comment by sohami** [comment]
> ❕ Gradle check result for [756e26a](https://github.com/opensearch-project/OpenSearch/commit/756e26acdd33f85b9c203dbcbd366fa01a308af8): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/39111/)
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Created tracking issue for new flaky test: https://github.com/opensearch-project/OpenSearch/issues/13797. Other one is known: https://github.com/opensearch-project/OpenSearch/issues/13307

**Review by sohami — APPROVED** [review]

**Pull request merged**